[torch.compile] Fine-grained CustomOp enabling mechanism

ProExpertProg opened the pull request:
Add support for fine-grained enabling of custom ops via the `VLLM_CUSTOM_OPS` environment variable.

This PR addresses the comment in #9058 about fine-grained enablement of custom ops.

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

ProExpertProg committed:
add custom op enabling mechanism (with test)

ProExpertProg committed:
Add a comment

ProExpertProg committed:
revert comment

youkaichao commented on vllm/envs.py:
```diff
@@ -205,7 +206,18 @@ def get_default_config_root():
+    # - 'none,op1,op2' to enable only op1 and op2
+    # By default, all custom ops are enabled when running without Inductor
+    # and disabled when running with Inductor (compile_level >= Inductor).
+    "VLLM_ENABLE_CUSTOM_OPS":
```
we can rename it to `VLLM_CUSTOM_OPS` , with the following semantic:

if `envs.VLLM_TORCH_COMPILE_LEVEL >= CompilationLevel.INDUCTOR` , all custom ops are disabled by default, and we only check `+rms_norm` etc to enable some of them.

if `envs.VLLM_TORCH_COMPILE_LEVEL < CompilationLevel.INDUCTOR` , all custom ops are enabled by default, and we only check `-rms_norm` etc to disable some of them.

ProExpertProg replied:
I assume we still want to support `all` or `none` though if someone wants to disable all custom ops without inductor or enable all with inductor?

youkaichao commented on vllm/model_executor/layers/activation.py:
```diff
@@ -58,7 +61,7 @@ class GeluAndMul(CustomOp):
 
     def __init__(self, approximate: str = "none"):
-        super().__init__()
+        super().__init__("gelu_and_mul")
```
please use `name=`

youkaichao commented on vllm/model_executor/layers/activation.py:
```diff
@@ -98,6 +101,9 @@ def extra_repr(self) -> str:
 class NewGELU(CustomOp):
 
+    def __init__(self):
+        super().__init__("gelu_new")
```
can we make it a class attribute? and there should be some registry in the `CustomOp` class, so that we can call some functions to list all custom ops.

youkaichao commented on vllm/model_executor/custom_op.py:
```diff
@@ -69,3 +76,61 @@ def dispatch_forward(self):
+        # On by default if VLLM_TORCH_COMPILE_LEVEL < CompilationLevel.INDUCTOR
+        default_on = envs.VLLM_TORCH_COMPILE_LEVEL < CompilationLevel.INDUCTOR
+
+        # override the default if 'all' or 'none' is specified
+        default_on = default_on and not bool(use_none) or bool(use_all)
+        enabled_ops, disabled_ops = set(), set()
+
+        for op in env_ops:
+            if op == "all" or op == "none":
+                continue
+
+            assert re.match(r"^-?[a-z0-9_]+$",
+                            op), f"Invalid custom op name: '{op}'"
+
+            if op.startswith("-"):
+                assert op[1:] not in {"all", "none"}, \
+                    f"Cannot disable all or none: '{op}'"
+                disabled_ops.add(op[1:])
+            else:
+                enabled_ops.add(op)
```
this is too complicated. we can just use if `+rms_norm` in `envs.VLLM_CUSTOM_OPS`

ProExpertProg committed:
PR comments:
- removed complex initialization
- name -> class property

ProExpertProg committed:
Reformat test

ProExpertProg committed:
Add custom op registry

bnellnm commented on vllm/model_executor/custom_op.py:
```diff
@@ -69,3 +75,40 @@ def dispatch_forward(self):
+
+    @classmethod
+    def enabled(cls) -> bool:
+        enabled = f"+{cls.name}" in envs.VLLM_CUSTOM_OPS
```
would it be simpler to just use the python class name and not have to register anything?

ProExpertProg replied:
The last commit adds a registry of custom ops which requires the registration anyway. So I'm inclined to keep it this way - but we could change it to class names to avoid different capitalization?

bnellnm replied:
What happens if someone forgets to register their CustomOp?  Will it use the global enable/disable settings?  

Also, is there a way to avoid the manual name setting in `FlippedSiluAndMul`?  Can you not use register on that class too?

ProExpertProg replied:
I assume it will fail as it won't find `cls.name` but I can improve the error message here.

ProExpertProg replied:
FlippedSiluAndMul is an oversight - it shouldn't be registered and should just use the SiluAndMul settings - but I'll verify that's how it works.

bnellnm commented:
LGTM.  My only question/suggestion was to use the python class names for simplicity (with the module(s) stripped I suppose).

youkaichao commented on vllm/model_executor/custom_op.py:
```diff
@@ -56,7 +62,7 @@ def dispatch_forward(self):
         # specific backend. Currently, we do not support dynamic dispatching.
 
-        if envs.VLLM_TORCH_COMPILE_LEVEL >= CompilationLevel.INDUCTOR:
+        if not self.enabled():
```
```suggestion
-        if not self.enabled():
+        enabled = self.enabled()
+        name = self.__class__.name
+        logger.debug("custom op %s %s", name,
+                     "enabled" if enabled else "disabled")
+        if not enabled:
```

add some logging here, so that we know which ops are used for the current model we run

youkaichao reviewed (approved):
thanks for the great efforts!

tlrmchlsmth reviewed (approved):
lgtm

ProExpertProg commented on vllm/model_executor/models/ultravox.py:
```diff
@@ -231,6 +231,7 @@ def forward(self, audio_embeds: torch.Tensor) -> torch.Tensor:
 
 class FlippedSiluAndMul(SiluAndMul):
     """Ultravox is trained with SwiGLU with flipped halves."""
+    name = "flipped_silu_and_mul"
```
Removed

ProExpertProg replied:
Sorry, to clarify: this is a subclass of another CustomOp and it's in a model definition so I figured it didn't need to be registered. It can be enabled/disabled using the parent name (`"silu_and_mul"`), but let me know if you think we should register it

ProExpertProg committed:
PR comments:
- warning when not registered
- log when enabled
- test subclass behavior
- remove name from model FlippedSiluAndMul

ProExpertProg committed:
Fix comment in envs

ProExpertProg committed:
Improve custom_op comment

ProExpertProg committed:
Merge branch 'refs/heads/main' into luka/custom_ops_env

# Conflicts:
#	vllm/envs.py
#	vllm/model_executor/layers/activation.py

youkaichao commented:
> 
> [2024-10-17T14:46:39Z] DEBUG 10-17 07:46:39 custom_op.py:69] custom op gelu_fast enabled
> --
>   | [2024-10-17T14:46:39Z] DEBUG 10-17 07:46:39 custom_op.py:69] custom op gelu_new enabled
>   | [2024-10-17T14:46:39Z] DEBUG 10-17 07:46:39 custom_op.py:69] custom op relu2 enabled
>   | [2024-10-17T14:46:39Z] DEBUG 10-17 07:46:39 custom_op.py:69] custom op quick_gelu enabled
> 

this is caused by:

> _ACTIVATION_REGISTRY = {
>     "gelu": nn.GELU(),
>     "gelu_fast": FastGELU(),
>     "gelu_new": NewGELU(),
>     "gelu_pytorch_tanh": nn.GELU(approximate="tanh"),
>     "relu": nn.ReLU(),
>     "relu2": ReLUSquaredActivation(),
>     "quick_gelu": QuickGELU(),
> }
> 

@ProExpertProg we should make them lazy init, otherwise the logging information is confusing.

ProExpertProg commented:
> @ProExpertProg we should make them lazy init, otherwise the logging information is confusing.

You're saying we should lazy-init the activation registry right?

youkaichao commented:
yes, we should:

```python
_ACTIVATION_REGISTRY = {
"gelu": None,
}
```

and only `_ACTIVATION_REGISTRY["gelu"] = nn.GELU()` after the first call of `get_act_fn("gelu")`

ProExpertProg committed:
Lazy init for activations

ProExpertProg commented:
I added a LazyDict implementation instead, let me know what you think

tlrmchlsmth commented on vllm/utils.py:
```diff
@@ -1442,3 +1443,24 @@ def dec(self, num=1):
+
+
+# Adapted from: https://stackoverflow.com/a/47212782/5082708
+class LazyDict(Mapping, Generic[T]):
```
cool
